Pragma-driven subroutine inlining and associated utilities

mlange05 opened the pull request:
The primary purpose of this PR is to introduce pragma-driven source inlining for Suboutine where `CallStatement` nodes annotated with the `!$loki inline` pragma are replaced by the body of the called routine, with resolved argument mappings. As I developed this on the EC-physics regression suite, I needed to include a few additional utilities that deal with the fallout of the resolution of optional arguments (simplistic logic evaluation in `simplify` and primitive dead code elimination), but these could easily be factored into a separate PR if that is desirable (@reuterbal, please shout!). 

Otherwise, the PR is primarily a refactoring of the procedure inlining utility, where the remapping of a call to the called body is now completely separated from the dealing with local variable declarations (in `inline_subroutine_calls`). This allows us to reuse these entry points between `inline_marked_subroutines` and `inline_internal_procedures`.

The rest of the changes and fixes some more detail:
- Separation of call remapping from declaration/scope management in procedure inlining
- Renaming of `inline_member_subroutines` to the more accurate "internal" and "procedure" (no change in capability though!)
- Dealing with unresolved `optional` arguments after call remapping (resolve the `present(n)` clause, depending on the presence of `n`, to either `.true.` or `.false.`)
-  Simplistic dead code elimination that resolves explicitly unreachable branches of conditionals (eg. from resolving `if (present(arg)) <do_something_with_arg>`. This then avoids dealing with remnant uses of the optional args, in case they were not given to an inlined call.
- Extension of the above dead code elimination by resolving simple constant expressions to logical values, eg. `1 == 6 .or. 2 == 4` to simply `.false.`
- A small fix to vector section-derivation that fixes a corner case of `!$loki separator` prgamas sitting deep in a nested control flow (possibly under tested :hand_over_mouth: )

MichaelSt98 commented on tests/test_transform_inline.py:
```diff
@@ -607,3 +609,145 @@ def test_inline_member_routines_with_associate(frontend):
+  subroutine add_a_to_b(a, b, n)
+    real(kind=8), intent(inout) :: a(:), b(:)
+    integer, intent(in) :: n
+
```
missing `integer :: i`, which changes the inlined code as well 

```fortran
  DO add_a_to_b_i=1,add_a_to_b_n
    a(add_a_to_b_i) = a(add_a_to_b_i) + b(add_a_to_b_i)
  END DO
```

Thus still works and is valid, just the question whether this should be fixed for possible future changes?

reuterbal replied:
Yes, original Fortran should be valid (technically it is, because there's also no `implicit none`)

mlange05 replied:
Ahh, this is what the `allowed_aliases` is for; I've amended the test to use this and retrofitted the `implicit none`.

MichaelSt98 reviewed (approved):
Really nice. Looking forward to use this 👍

reuterbal commented on tests/test_transform_inline.py:
```diff
@@ -607,3 +609,145 @@ def test_inline_member_routines_with_associate(frontend):
     assert len(assocs) == 2
+
+
+@pytest.mark.parametrize('frontend', available_frontends(xfail=[(OMNI, 'No header information in test')]))
```
Why should this not work for OMNI? If you parse the module first, the missing xmod is auto-generated and the driver should parse successfully.

mlange05 replied:
Correct, wasn't thinking! Many thanks for picking this out!

reuterbal commented on tests/test_transform_inline.py:
```diff
@@ -607,3 +609,145 @@ def test_inline_member_routines_with_associate(frontend):
+    assert calls[0].arguments == ('b(i)',)
+
+
+@pytest.mark.parametrize('frontend', available_frontends(xfail=[(OMNI, 'No header information in test')]))
```
Same comment as in the previous test

reuterbal commented on tests/test_transform_dead_code.py:
```diff
@@ -0,0 +1,52 @@
+    end if
+
+    if (2 == 2) then
+      b = b + a
```
Would you mind adding a second test with conditionals with `elseif` branches that exercise the permutations of options which paths are to be eliminated (if, elseif, else, if&elseif, elseif&else)? I expect this to largely work as expected without further work thanks to the multi-body conditionals being converted to if/else blocks. But the issue discovered by Paddy makes me think we may lack an update of the corresponding `is_elseif` flag.

mlange05 replied:
Yes, I've added two small tests for nested conditionals - one to the base test that gets fully resolved, and one additional test for partially resolved nested conditionals. The bug you feared never materialised though. :wink:

reuterbal replied:
Yeah, because you're leaving out the case that the conditional remains but the elseif branch is eliminated. See below for a suggestion how to amend the test

reuterbal commented on scripts/loki_transform.py:
```diff
@@ -107,15 +107,19 @@ def cli(debug):
               help='Replace array arguments passed as scalars with arrays.')
 @click.option('--derive-argument-array-shape/--no-derive-argument-array-shape', default=False,
               help="Recursively derive explicit shape dimension for argument arrays")
+@click.option('--eliminate_dead_code/--no-eliminate_dead_code', default=True,
```
I'd prefer to use hyphen notation instead of mixing with underscores.

mlange05 replied:
Fully agree. Copy-paste error ... fixed now.

reuterbal commented on loki/transform/transform_inline.py:
```diff
@@ -229,88 +227,140 @@ def _map_unbound_dims(var, val):
+        if v.name in parent_variables and v.name.lower() not in callee._dummies
     )
+    # Filter out allowed aliases to prevent suffixing
+    duplicates = tuple(v for v in duplicates if v.name not in allowed_aliases)
```
Do we need to be careful here with regard to case-sensitivity, since comparing string `v.name`?

mlange05 replied:
Aha, yes. But a better fix is to use `v.symbol` as our base symbols do indeed inherit case-insensitive equivalence. :grin: I've added this to the base test.

reuterbal commented on loki/transform/transform_dead_code.py:
```diff
@@ -0,0 +1,66 @@
+        if condition == 'False':
+            return else_body
+
+        return self._rebuild(o, tuple((condition,) + (body,) + (else_body,)))
```
I believe here we need a check for the case that [`has_elseif`](https://github.com/ecmwf-ifs/loki/blob/37d0033ea165985fb35cec2273067024af8e1e11/loki/ir.py#L566C5-L566C15) is True, whether the `else_body` still starts with a conditional as that part could have been resolved.

mlange05 replied:
This one is curious, as we actually do not need to check for it. The reasoning is that if we reach this line, none of the sub-branches have been resolved, so the current value of `has_elseif` is still valid. I've added tests to ensure this is covered though.

reuterbal reviewed (changes requested):
Really nice and powerful addition and I'm pleasantly surprised that we can do something as reasonably advanced as the dead code elimination. Many thanks for this!

I left a few comments throughout. No showstoppers but I suspect we may run into problems with if-elseif-else style conditionals when rebuilding the nodes. Also, not quite clear on why the tests shouldn't work with OMNI.

skarppinen commented:
Hello @mlange05, is there a particular reason to not add `allowed_aliases` for `inline_member_procedures` as well?

skarppinen commented:
Here is also a reproducer for a small bug in `inline_internal_procedures`:

This code
```
import loki
from loki.transform import inline_internal_procedures
fcode = """
    subroutine outer() 
        integer :: z
        integer :: jlon
        z = 0
        jlon = 0

        !$loki inline
        call inner(z)

        jlon = z + 4
        contains
        subroutine inner(z)
            integer, intent(inout) :: z
            integer :: jlon, jg ! This declaration with jlon and jg together causes the issue!
            jlon = 1
            jg = 2
            z = jlon + jg
        end subroutine inner
    end subroutine outer
"""
src = loki.Sourcefile.from_source(fcode)
routine = src.routines[0]
inline_internal_procedures(routine, "jlon")
```
produces for `routine` the Fortran source:
```
SUBROUTINE outer ()
  INTEGER :: z
  INTEGER :: jlon
  z = 0
  jlon = 0

!$loki inline
  ! [Loki] inlined child subroutine: inner
  ! =========================================
  jlon = 1
  jg = 2
  z = jlon + jg
  ! =========================================

  jlon = z + 4
  CONTAINS
END SUBROUTINE outer
```
where `jg` is never declared. This happens, because `jlon` and `jg` are declared on the same line in the inner routine. If they are declared each on their own lines, this issue disappears. Thus, I suspect that a call to `single_variable_declaration` should be made at some point.

mlange05 commented:
> Here is also a reproducer for a small bug in `inline_internal_procedures`:
> 
> This code
> 
> ```
> import loki
> from loki.transform import inline_internal_procedures
> fcode = """
>     subroutine outer() 
>         integer :: z
>         integer :: jlon
>         z = 0
>         jlon = 0
> 
>         !$loki inline
>         call inner(z)
> 
>         jlon = z + 4
>         contains
>         subroutine inner(z)
>             integer, intent(inout) :: z
>             integer :: jlon, jg ! This declaration with jlon and jg together causes the issue!
>             jlon = 1
>             jg = 2
>             z = jlon + jg
>         end subroutine inner
>     end subroutine outer
> """
> src = loki.Sourcefile.from_source(fcode)
> routine = src.routines[0]
> inline_internal_procedures(routine, "jlon")
> ```
> 
> produces for `routine` the Fortran source:
> 
> ```
> SUBROUTINE outer ()
>   INTEGER :: z
>   INTEGER :: jlon
>   z = 0
>   jlon = 0
> 
> !$loki inline
>   ! [Loki] inlined child subroutine: inner
>   ! =========================================
>   jlon = 1
>   jg = 2
>   z = jlon + jg
>   ! =========================================
> 
>   jlon = z + 4
>   CONTAINS
> END SUBROUTINE outer
> ```
> 
> where `jg` is never declared. This happens, because `jlon` and `jg` are declared on the same line in the inner routine. If they are declared each on their own lines, this issue disappears. Thus, I suspect that a call to `single_variable_declaration` should be made at some point.

Hi @skarppinen, many thanks for flagging this and providing the test. The reason this was not added was indeed just me being focused on the problem at hand, but I've now extended this option to the internal procedure inliner too.

Secondly, the test you provided indeed then causes a bug, and the fix follows closely to what you suggest (explicitly separate the "allowed alias" declarations into single-variable declarations using the utility. Please have a look and see of this fixes your issue.

mlange05 commented:
@reuterbal and @skarppinen many thanks for the thorough review and valuable suggestions. I think I've addressed all points, so this is ready for another look.

reuterbal commented on tests/test_transform_dead_code.py:
```diff
@@ -0,0 +1,101 @@
+    else
+      a = a + 6.0
+    end if
+
+  end if
+end subroutine test_dead_code_conditional
+"""
+    routine = Subroutine.from_source(fcode, frontend=frontend)
+    # Please note that nested conditionals (elseif) counts as two
+    assert len(FindNodes(Conditional).visit(routine.body)) == 5
+    assert len(FindNodes(Assignment).visit(routine.body)) == 7
+
+    dead_code_elimination(routine)
+
+    conditionals = FindNodes(Conditional).visit(routine.body)
+    assert len(conditionals) == 1
+    assert conditionals[0].condition == 'flag'
+    assigns = FindNodes(Assignment).visit(routine.body)
+    assert len(assigns) == 3
+    assert assigns[0].lhs == 'b' and assigns[0].rhs == 'b + 2.0'
+    assert assigns[1].lhs == 'b' and assigns[1].rhs == 'b + a'
+    assert assigns[2].lhs == 'a' and assigns[2].rhs == 'a + b'
```
```suggestion

    if (a > 2.0) then
      a = a + 5.0
    elseif (2 == 3) then
      a = a + 3.0
    else
      a = a + 1.0
    endif
  end if
end subroutine test_dead_code_conditional
"""
    routine = Subroutine.from_source(fcode, frontend=frontend)
    # Please note that nested conditionals (elseif) counts as two
    assert len(FindNodes(Conditional).visit(routine.body)) == 7
    assert len(FindNodes(Assignment).visit(routine.body)) == 10

    dead_code_elimination(routine)

    conditionals = FindNodes(Conditional).visit(routine.body)
    assert len(conditionals) == 2
    assert conditionals[0].condition == 'flag'
    assert not conditionals[0].has_elseif
    assert conditionals[1].condition == 'a > 2.0'
    assert not conditionals[1].has_elseif
    assigns = FindNodes(Assignment).visit(routine.body)
    assert len(assigns) == 5
    assert assigns[0].lhs == 'b' and assigns[0].rhs == 'b + 2.0'
    assert assigns[1].lhs == 'b' and assigns[1].rhs == 'b + a'
    assert assigns[2].lhs == 'a' and assigns[2].rhs == 'a + b'
    assert assigns[3].lhs == 'a' and assigns[3].rhs == 'a + 5.0'
    assert assigns[4].lhs == 'a' and assigns[4].rhs == 'a + 1.0'
```

This fails as expected, because the `has_elseif` flag remains for the last conditional although the conditional nested into the else branch has vanished:

```
self = Conditional::

    def __post_init__(self):
        super().__post_init__()
        assert self.condition is not None

        if self.body is not None:
            assert isinstance(self.body, tuple)
            assert all(isinstance(c, Node) for c in self.body)  # pylint: disable=not-an-iterable

        if self.has_elseif:
            assert len(self.else_body) == 1
>           assert isinstance(self.else_body[0], Conditional)  # pylint: disable=unsubscriptable-object
E           AssertionError

loki/ir.py:610: AssertionError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[36] > /etc/ecmwf/nfs/dh1_home_a/nabr/loki/naml-routine-inline-pragma/loki/ir.py(610)__post_init__()
-> assert isinstance(self.else_body[0], Conditional)  # pylint: disable=unsubscriptable-object
   5 frames hidden (try 'help hidden_frames')
```

mlange05 replied:
Aha, I knew I forgot something! :wink: Thanks for pointing this out, I've adjusted the test now accordingly and fixed the bug.

reuterbal reviewed (changes requested):
Many thanks, looks all good except for the `has_elseif` situation: That one I think isn't covered with the tests you added - I've proposed an addition which fails for me in the way I expected it to. Would you mind taking a look at that?

reuterbal commented on loki/transform/transform_dead_code.py:
```diff
@@ -63,4 +64,5 @@ def visit_Conditional(self, o, **kwargs):
             return else_body
 
-        return self._rebuild(o, tuple((condition,) + (body,) + (else_body,)))
+        has_elseif = isinstance(else_body, Conditional)
```
Hm, not sure this works as intended. The `else_body` should always be a tuple, not an individual node - and it might misclassify genuinely nested (albeit equivalent) conditionals as `has_elseif`.
I think something like this might work:
```suggestion
        has_elseif = o.has_elseif and else_body and isinstance(else_body[0], Conditional)
```

mlange05 replied:
Gotcha! Many thanks, fixed now (with additional testing).

skarppinen commented on transformations/transformations/single_column_coalesced.py:
```diff
@@ -296,14 +312,23 @@ def process_kernel(self, routine):
-        # Perform full source-inlining for member subroutines if so requested
+        # Perform full source-inlining for member subroutines
         if self.inline_members:
             inline_member_procedures(routine)
```
Should this call still be changed to `inline_member_procedures(routine, allowed_aliases=(self.horizontal.index,))`?

mlange05 replied:
Ah, good catch. I have no means of testing this, but seems easy enough, so added as suggested.

reuterbal reviewed (approved):
Thanks, that looks great now. Ready to merge! :shipit:
